src/04-extensions.js

```diff
@@ -0,0 +1,212 @@
+const exampleDinosaurData = require("../data/dinosaurs");
+const exampleTicketData = require("../data/tickets");
+const exampleRoomData = require("../data/rooms");
+
+/*
+Write a function that will organize all of the dinosaurs by their period key. 
+The function should return an object where the keys are the period values, and 
+the values are an array of IDs.
+*/
+
+/* Expected Result
+  {
+    "Late Jurassic": [
+      "YLtkN9R37",
+      "GGvO1X9Zeh",
+      "BFjjLjea-O",
+      "iOVNUcv-ww",
+      "V53DvdhV2A",
+    ],
+    "Late Cretaceous": [
+      "WHQcpcOj0G",
+      "GKl035EYKN",
+      "2GglUqKT0G",
+      "Pr6kc4Q_Xf",
+      "ft5Gs5izdq",
+      "wuL4ddBinQ",
+    ],
+    // ...
+  }
+*/
+
+function organizeDinosaursByPeriod(dinos, key) {
+  //input an array of objects
+  //output an object where key represents period and value reps an array of dinoIDs
+
+  //create helper function to camelCase
+  function camelCaseKey(keyString) {
+    //create an array that contains all valid keys
+    const validKeys = ["name", "pronunciation", "meaningOfName", "diet", "lengthInMeters", "mya", "info"];
+
+    //iterate through validKeys and if the keyString is included in array then assign it
+    for (let i = 0; i < validKeys.length; i++) {
+      if (validKeys[i].toLowerCase() === keyString.toLowerCase()) {
+        keyString = validKeys[i];
+        break;
+      }
+    }
+
+    return keyString;
+  }
+
+  //if key is undefined assign it the string dinosaurId
+  if (key === undefined) key = "dinosaurId";
+  else key = camelCaseKey(key);
+  //edgecase - all dinos share same keys, so check whether key exists in object
+  if (!dinos[0][key] || key === "period") return "please provide a valid key: (name,pronunciation,meaningOfName,diet,lengthInMeters,mya, or info)";
+
+  //initialie an empty object
+  const result = {};
+
+  //iterate through dinosaurs
+  dinos.forEach((dino) => {
+    //if period already exists in object, push dinoID to object at key
+    if (result[dino.period]) result[dino.period].push(dino[key]);
+    //otherwise create key in object and assign it array with current dinoID in it
+    else result[dino.period] = [dino[key]];
+  });
+
+  return result;
+}
+
+const result = organizeDinosaursByPeriod(exampleDinosaurData, "mya");
+// console.log(result);
+
+//--SPACING
+
+/*
+Write a function that takes in an array of strings, where each string is a roomId. Assume that someone is attempting to go from room to 
+room, starting with the room at index 0 and then preceding to the room at index 1, and so on.
+Return true if the path the person is trying to follow is possible. Otherwise, return false.
+*/
+
+const valid = ["aIA6tevTne", "A6QaYdyKra", "L72moIRcrX", "0eNtkY5WoA"];
+const invalid = ["aIA6tevTne", "A6QaYdyKra", "L72moIRcrX", "dpQnu5wgaN"];
+
+function validateMuseumPath(selectedPath, showPath) {
```

Comment:
I think this function isn't quite working as intended. For example, here's some output I get when running it:
```js
validateMuseumPath([ "aIA6tevTne", "A6QaYdyKra", "L72moIRcrX" ], true);
//> 'Coat Check Room -> Ticket Center -> Ellis Family Hall '
validateMuseumPath([ "aIA6tevTne", "A6QaYdyKra", "L72moIRcrX", "A6QaYdyKra" ], true)
//> 'Coat Check Room -> Ticket Center -> Kit Hopkins Education Wing -> Haley Hall '
```
This seems wrong to me because:
1. The first three IDs are the same in both calls, but the room name shown is different.
2. The last room in the last call, `"A6QaYdyKra"`, is returning back to the ticket center, but it has a completely different name?